dbms/src/Flash/Mpp/ReceivedMessageQueue.cpp

```diff
@@ -0,0 +1,150 @@
+                auto pop_result [[maybe_unused]] = grpc_recv_queue->tryPop(original_msg);
+                /// if there is no remaining consumer, then pop it from original queue, the message must stay in the queue before the pop
+                /// so even use tryPop, the result must not be empty
+                assert(pop_result != MPMCQueueResult::EMPTY);
+                if (original_msg != nullptr)
+                    RUNTIME_CHECK_MSG(*original_msg->remaining_consumers == 0, "Fine grained receiver pop a message that is not full consumed, remaining consumer: {}", *original_msg->remaining_consumers);
+            }
+        }
+    }
+    else
+    {
+        if constexpr (need_wait)
+        {
+            res = grpc_recv_queue->pop(recv_msg);
+        }
+        else
+        {
+            res = grpc_recv_queue->tryPop(recv_msg);
+        }
+    }
+    return {res, recv_msg};
+}
+
+template <bool is_force>
+bool ReceivedMessageQueue::pushToMessageChannel(ReceivedMessagePtr & received_message, ReceiverMode mode)
+{
+    bool success = false;
+    if constexpr (is_force)
+        success = msg_channel->forcePush(received_message) == MPMCQueueResult::OK;
+    else
+        success = msg_channel->push(received_message) == MPMCQueueResult::OK;
+
+    injectFailPointReceiverPushFail(success, mode);
+    return success;
+}
+
+GRPCReceiveQueueRes ReceivedMessageQueue::pushToGRPCReceiveQueue(ReceivedMessagePtr & received_message)
+{
+    auto res = grpc_recv_queue->push(received_message);
+    fiu_do_on(FailPoints::random_receiver_async_msg_push_failure_failpoint, res = GRPCReceiveQueueRes::CANCELLED);
+    return res;
+}
+
+ReceivedMessageQueue::ReceivedMessageQueue(
+    const AsyncRequestHandlerWaitQueuePtr & conn_wait_queue,
+    const LoggerPtr & log_,
+    size_t max_buffer_size,
+    bool enable_fine_grained,
+    size_t fine_grained_channel_size_)
+    : fine_grained_channel_size(enable_fine_grained ? fine_grained_channel_size_ : 0)
+    , log(log_)
+{
+    if (enable_fine_grained)
+    {
+        for (size_t i = 0; i < fine_grained_channel_size; ++i)
+            /// these are unbounded queues
+            msg_channels_for_fine_grained_shuffle.push_back(std::make_shared<LooseBoundedMPMCQueue<ReceivedMessagePtr>>(std::numeric_limits<size_t>::max()));
+        /// use pushcallback to make sure that the order of messages in msg_channels_for_fine_grained_shuffle is exactly the same as it in msg_channel,
+        /// because pop from msg_channel rely on this assumption. An alternative is to make msg_channel a set/map of messages for fine grained shuffle, but
+        /// it need many more changes
+        msg_channel = std::make_shared<LooseBoundedMPMCQueue<ReceivedMessagePtr>>(max_buffer_size, [this](const ReceivedMessagePtr & element) {
+            for (size_t i = 0; i < fine_grained_channel_size; ++i)
+            {
+                if (msg_channels_for_fine_grained_shuffle[i]->tryPush(element) != MPMCQueueResult::OK)
```

Comment:
use `forcePush`?

Reply:
Since the queue is unbounded, I think both `forcePush` and `tryPush` is ok.